contracts/StakingRewards.sol

```diff
@@ -187,89 +225,91 @@ contract StakingRewards is IStakingRewards, ManagedContract {
+    /// @param guardian is the delegator's guardian prior to the change
+    /// @param guardianDelegatedStake is the delegated stake of the delegator's guardian prior to the change
+    /// @param delegator is the delegator about to change delegation state
+    /// @param delegatorStake is the stake of the delegator
+    /// @param nextGuardian is the delegator's guardian after to the change
+    /// @param nextGuardianDelegatedStake is the delegated stake of the delegator's guardian after to the change
+    function delegationWillChange(address guardian, uint256 guardianDelegatedStake, address delegator, uint256 delegatorStake, address nextGuardian, uint256 nextGuardianDelegatedStake) external override onlyWhenActive onlyDelegationsContract {
+        Settings memory _settings = settings;
+        (bool inCommittee, uint256 weight, , uint256 totalCommitteeWeight) = committeeContract.getMemberInfo(guardian);
+        
+        StakingRewardsState memory _stakingRewardsState = _updateStakingRewardsState(totalCommitteeWeight, _settings);
+        GuardianStakingRewards memory guardianStakingRewards = _updateGuardianStakingRewards(guardian, inCommittee, inCommittee, weight, guardianDelegatedStake, _stakingRewardsState, _settings);
+        _updateDelegatorStakingRewards(delegator, delegatorStake, guardian, guardianStakingRewards);
 
-        for (uint i = 0; i < addrs.length; i++) {
-            guardiansStakingRewards[addrs[i]].balance = guardiansStakingRewards[addrs[i]].balance.add(migratedGuardianStakingRewards[i]);
-            delegatorsStakingRewards[addrs[i]].balance = delegatorsStakingRewards[addrs[i]].balance.add(migratedDelegatorStakingRewards[i]);
-            emit StakingRewardsBalanceMigrationAccepted(msg.sender, addrs[i], migratedGuardianStakingRewards[i], migratedDelegatorStakingRewards[i]);
+        if (nextGuardian != guardian) {
+            (inCommittee, weight, , totalCommitteeWeight) = committeeContract.getMemberInfo(nextGuardian);
+            GuardianStakingRewards memory nextGuardianStakingRewards = _updateGuardianStakingRewards(nextGuardian, inCommittee, inCommittee, weight, nextGuardianDelegatedStake, _stakingRewardsState, _settings);
+            delegatorsStakingRewards[delegator].lastDelegatorRewardsPerToken = nextGuardianStakingRewards.delegatorRewardsPerToken;
         }
     }
 
-    function emergencyWithdraw(address erc20) external override onlyMigrationManager {
-        IERC20 _token = IERC20(erc20);
-        emit EmergencyWithdrawal(msg.sender, address(_token));
-        require(_token.transfer(msg.sender, _token.balanceOf(address(this))), "StakingRewards::emergencyWithdraw - transfer failed");
-    }
+    /*
+    * Governance functions
+    */
 
+    /// Activates staking rewards allocation
+    /// @dev governance function called only by the initialization admin
+    /// @dev On migrations, startTime should be set as the previous contract deactivation time.
```

Comment:
reads from the previous contract?

Reply:
May be useful also in non-migration scenarios. For example in order to adjust after a client rate changed